pkg/reconciler/rolloutorchestrator/strategies/strategy.go

```diff
@@ -0,0 +1,292 @@
+func (s *ScaleUpStep) Execute(ctx context.Context, ro *v1.RolloutOrchestrator, revScalingUp, _ map[string]*v1.TargetRevision) error {
+	// Create or update the StagePodAutoscaler for the revision scale up
+	for _, revUp := range revScalingUp {
+		if _, err := s.CreateOrUpdateSPARev(ctx, ro, revUp, true, UpdateSPAForRevUp); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *ScaleUpStep) Verify(ctx context.Context, ro *v1.RolloutOrchestrator, revScalingUp, revScalingDown map[string]*v1.TargetRevision,
+	_ func(interface{}, time.Duration)) (bool, error) {
+	for _, revUp := range revScalingUp {
+		spa, err := s.StagePodAutoscalerLister.StagePodAutoscalers(ro.Namespace).Get(revUp.RevisionName)
+		if err != nil {
+			return false, err
+		}
+
+		// spa.IsStageScaleInReady() returns true, as long as both DesireScale and ActualScale are available.
+		if !spa.IsStageScaleInReady() || !IsStageScaleUpReady(spa, revUp) {
+			// Create the stage pod autoscaler with the new maxScale set to
+			// maxScale defined in the revision traffic, because scale up phase is not over, we cannot
+			// scale down the old revision.
+			// Create or update the stagePodAutoscaler for the revision to be scaled down, even if the scaling up
+			// phase is not over.
+			for _, valDown := range revScalingDown {
+				if _, err = s.CreateOrUpdateSPARev(ctx, ro, valDown, false, UpdateSPAForRevDown); err != nil {
+					return false, err
+				}
+			}
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (s *ScaleUpStep) ModifyStatus(ro *v1.RolloutOrchestrator) {
+	ro.Status.MarkStageRevisionScaleUpReady()
+}
+
+// The ScaleUpStep struct is responsible for scaling down the pods for the old revisions.
+type ScaleDownStep struct {
+	BaseScaleStep
+}
+
+func (s *ScaleDownStep) Execute(ctx context.Context, ro *v1.RolloutOrchestrator, revScalingUp, revScalingDown map[string]*v1.TargetRevision) error {
+	// In the resourceUtil mode, we start to scale down the old revisions, but we need to create the SPAs for the new revisions,
+	// because Knative Serving launches the new revisions first. We need to leverage the SPAs to control the number of
+	// the pods for the new revisions.
+	// In the normal mode, we will always get the SPA for the new revision. The creation of the SPA will not be called.
+	for _, revUp := range revScalingUp {
+		_, err := s.StagePodAutoscalerLister.StagePodAutoscalers(ro.Namespace).Get(revUp.RevisionName)
+		if apierrs.IsNotFound(err) {
+			spa := &v1.StagePodAutoscaler{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      revUp.RevisionName,
+					Namespace: ro.Namespace,
+					Labels: map[string]string{
+						serving.RevisionLabelKey: revUp.RevisionName,
+						serving.ServiceLabelKey:  ro.Name,
+					},
+					OwnerReferences: []metav1.OwnerReference{
+						*kmeta.NewControllerRef(ro),
+					},
+				},
+				Spec: v1.StagePodAutoscalerSpec{
+					// If the SPA for the new revisions are created, start with 0 for the MinScale, and 1 for MaxScale.
```

Comment:
The comment is wrong as MaxScale is setting to 1, can we add an explanation why it is set this way?

Reply:
Yes, I will add the explanations.